drivers/pwm/pwm_mspm0.c

```diff
@@ -0,0 +1,168 @@
+
+#define REG_TIMA0            (GPTIMER_Regs *)0x40860000
+#define REG_TIMA1            (GPTIMER_Regs *)0x40862000
+#define IS_TIMA(tim)         (tim == REG_TIMA0 || tim == REG_TIMA1)
+// from the datasheet, only TIMA0 has 4 capture compare channels
+#define HAS_FOUR_CC(tim)     (tim == REG_TIMA0)
+#define MAX_ALLOWED_CHANNELS_TIMX         2
+#define MAX_ALLOWED_CHANNELS_WITH_FOUR_CC 4
+
+LOG_MODULE_REGISTER(pwm_mspm0, CONFIG_PWM_LOG_LEVEL);
+
+struct pwm_mspm0_config {
+	GPTIMER_Regs *timer;
+	const struct pinctrl_dev_config *pinctrl;
+	const uint32_t frequency;
+	const DL_Timer_ClockConfig clock_cfg;
+	const DL_TIMER_PWM_MODE pwm_mode;
+};
+
+static inline int pwm_mspm0_get_cc_direction(int cc_index, int *direction)
+{
+	switch (cc_index) {
+	case DL_TIMER_CC_0_INDEX:
+		*direction = DL_TIMER_CC0_OUTPUT;
+		return 0;
+	case DL_TIMER_CC_1_INDEX:
+		*direction = DL_TIMER_CC1_OUTPUT;
+		return 0;
+	case DL_TIMER_CC_2_INDEX:
+		*direction = DL_TIMER_CC2_OUTPUT;
+		return 0;
+	case DL_TIMER_CC_3_INDEX:
+		*direction = DL_TIMER_CC3_OUTPUT;
+		return 0;
+	default:
+		LOG_DBG("Error, invalid CC index: %d", cc_index);
+		return -EINVAL;
+	};
+}
+
+static int pwm_mspm0_get_cycles_per_sec(const struct device *dev, uint32_t channel,
+					uint64_t *cycles)
+{
+	const struct pwm_mspm0_config *config = dev->config;
+
+	*cycles = config->frequency /
+		  ((config->clock_cfg.divideRatio + 1) * (config->clock_cfg.prescale + 1));
+
+	return 0;
+}
+
+static inline uint32_t get_polarity(pwm_flags_t flags)
+{
+	if ((flags & PWM_POLARITY_MASK) == PWM_POLARITY_NORMAL) {
+		return DL_TIMER_CC_OCTL_INV_OUT_DISABLED;
+	}
+
+	return DL_TIMER_CC_OCTL_INV_OUT_ENABLED;
+}
+
+static int pwm_mspm0_set_cycles(const struct device *dev, uint32_t channel, uint32_t period_cycles,
+				uint32_t pulse_cycles, pwm_flags_t flags)
+{
+	int direction;
+	const struct pwm_mspm0_config *cfg = dev->config;
+	uint8_t max_allowed_channels = HAS_FOUR_CC(cfg->timer) ? MAX_ALLOWED_CHANNELS_WITH_FOUR_CC
+							       : MAX_ALLOWED_CHANNELS_TIMX;
+
+	if (channel >= max_allowed_channels) {
+		LOG_ERR("Error, invalid channel: %d", channel);
+		return -EINVAL;
+	}
+
+	if (pulse_cycles == 0) {
+		DL_Timer_stopCounter(cfg->timer);
+		return 0;
+	}
+
+	if (IS_TIMA(cfg->timer)) {
+		const DL_TimerA_PWMConfig pwm_cfg = {
+			.pwmMode = cfg->pwm_mode,
+			.isTimerWithFourCC = HAS_FOUR_CC(cfg->timer),
+			.period = period_cycles,
+		};
+
+		DL_TimerA_initPWMMode(cfg->timer, (DL_TimerA_PWMConfig *)&pwm_cfg);
+	} else {
+		const DL_TimerG_PWMConfig pwm_cfg = {
+			.pwmMode = cfg->pwm_mode,
+			.period = period_cycles,
+		};
+
+		DL_TimerG_initPWMMode(cfg->timer, (DL_TimerG_PWMConfig *)&pwm_cfg);
+	}
+
+	DL_Timer_setCaptureCompareOutCtl(cfg->timer, DL_TIMER_CC_OCTL_INIT_VAL_HIGH,
+					 get_polarity(flags), DL_TIMER_CC_OCTL_SRC_FUNCVAL,
+					 channel);
+	int err = pwm_mspm0_get_cc_direction(channel, &direction);
+	if (err) {
+		DL_Timer_disablePower(cfg->timer);
+		return err;
```

Comment:
If this is hit, `err == -EINVAL` (from get_cc_direction). This also opposes the API which specifies to return -EINVAL specifically when pulse_cycles is greater than period_cycles